kong/clustering/rpc/queue.lua

```diff
@@ -0,0 +1,71 @@
+local semaphore = require("ngx.semaphore")
+local table_new = require("table.new")
+
+
+local assert = assert
+local setmetatable = setmetatable
+local math_min = math.min
+
+
+local _M = {}
+local _MT = { __index = _M, }
+
+
+local DEFAULT_QUEUE_LEN = 128
+
+
+function _M.new(max_len)
+  local self = {
+    semaphore = assert(semaphore.new()),
+    max = max_len,
+
+    elts = table_new(math_min(max_len, DEFAULT_QUEUE_LEN), 0),
+    first = 0,
+    last = -1,
+  }
+
+  return setmetatable(self, _MT)
+end
+
+
+function _M:push(item)
+  local last = self.last
+
+  if last - self.first + 1 >= self.max then
+    return nil, "queue overflow"
+  end
+
+  last = last + 1
+  self.last = last
+  self.elts[last] = item
+
+  self.semaphore:post()
+
+  return true
+end
+
+
+function _M:pop(timeout)
+  local ok, err = self.semaphore:wait(timeout)
+  if not ok then
+    if err == "timeout" then
```

Comment:
there is `utils.is_timeout` should it be used here?

Reply:
https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/semaphore.md#wait

>  or return nil and a string describing the error in case of failure (like "timeout").

So I am resolving this conversation

Reply:
@chronolaw which works just fine with:
https://github.com/Kong/kong/blob/feat/cp_dp_rpc/kong/clustering/rpc/utils.lua#L27

aka:
```lua
function _M.is_timeout(err)
  return err and string_sub(err, -7) == "timeout"
end
```

Perhaps that should also check direct `timeout` (but works without too:
```lua
function _M.is_timeout(err)
  return err and (err == "timeout" or  string_sub(err, -7) == "timeout")
end
```

@chobits, thus I am unresolving this.